azure/services/roleassignments/roleassignments.go

```diff
@@ -67,92 +68,85 @@ func (s *Service) Name() string {
 
 // Reconcile creates a role assignment.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
 	defer done()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+	log.V(2).Info("reconciling role assignment")
+
+	// Return early if there is no roleassignment spec to reconcile.
+	if len(s.Scope.RoleAssignmentSpecs(to.StringPtr(""))) < 1 {
+		log.V(2).Info("no role assignment spec to reconcile")
+		return nil
+	}
```

Comment:
This is a redundant call to role assignments spec that I had to make so that when we fetch principalID -- nil point exception does not occur. 

When there is not role assignment spec, while fetching the vm details, some of fields are nil and looks more simpler to return early in case there is no role assignment spec.

So this first call passes an empty principal ID and validates if role assignment spec is present and then below finds the principal ID and calls this spec function once again to finally get the correct role assignment specs. 
cc @CecileRobertMichon @Jont828

Reply:
I'd suggest adding another method to scope that just returns true/false if the machine/machinepool has system assigned identity for better readability.

Reply:
> I'd suggest adding another method to scope that just returns true/false if the machine/machinepool has system assigned identity for better readability.

+1

Reply:
Done